Update shared vpc config for Project Factory and Project module

simonebruzzechesse opened the pull request:
Update shared vpc config for project factory and project module for more granular Shared VPC configuration. Added the following attributes to shared_vpc_service_config variable:

- subnets_iam: for subnet level IAM binding for non API services identities
- host_project_iam: for granting network user role to identities other than API services

I applicable, I acknowledge that I have:
- [X] Read the [contributing guide](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md)
- [X] Ran `terraform fmt` on all modified files
- [X] Regenerated the relevant README.md files using [`tools/tfdoc.py`](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md#fabric-tools)
- [X] Made sure all relevant tests pass

ludoo commented:
wow you're fast :)

ludoo commented on modules/project/variables.tf:
```diff
@@ -302,10 +302,12 @@ variable "shared_vpc_service_config" {
-    service_identity_subnet_iam = optional(map(list(string)), {})
-    service_iam_grants          = optional(list(string), [])
+    host_project                 = string
+    host_project_iam             = optional(list(string), [])
```
I would restrict to only assign compute.networkUser, and rename to host_network_users

ludoo replied:
let's just rename to

- `network_users`
- `network_subnet_users`

simonebruzzechesse replied:
Updated variables and code!

ludoo commented on modules/project/variables.tf:
```diff
@@ -302,10 +302,12 @@ variable "shared_vpc_service_config" {
+    service_identity_iam         = optional(map(list(string)), {})
+    service_identity_subnets_iam = optional(map(list(string)), {})
+    service_iam_grants           = optional(list(string), [])
+    subnets_iam                  = optional(map(list(string)), {})
```
same as above, only network user role makes sense from service projects, and for service identities we already support more roles

simonebruzzechesse commented:
> wow you're fast :)

Not that fast.. I started workingon this yesterday! :)

ludoo commented on modules/project/shared-vpc.tf:
```diff
@@ -55,19 +55,37 @@ locals {
   ])
-  svpc_service_subnet_iam = {
-    for v in local._svpc_service_subnet_iam :
+  svpc_service_subnets_iam = {
```
`svpc_service_subnet_iam`

simonebruzzechesse replied:
Updated!

ludoo commented on modules/project/shared-vpc.tf:
```diff
@@ -55,19 +55,37 @@ locals {
   # normalize the service identity subnet IAM bindings
-  _svpc_service_subnet_iam = flatten([
-    for subnet, services in local._svpc.service_identity_subnet_iam : [
+  _svpc_service_subnets_iam = flatten([
```
`_svpc_service_subnet_iam`

simonebruzzechesse replied:
Updated!

ludoo commented on modules/project/variables.tf:
```diff
@@ -302,10 +302,12 @@ variable "shared_vpc_service_config" {
+    host_project                 = string
+    host_project_iam             = optional(list(string), [])
+    service_identity_iam         = optional(map(list(string)), {})
+    service_identity_subnets_iam = optional(map(list(string)), {})
```
`service_identity_subnet_iam`

simonebruzzechesse replied:
Updated!

ludoo commented on modules/project/variables.tf:
```diff
@@ -302,10 +302,12 @@ variable "shared_vpc_service_config" {
-    service_identity_subnet_iam = optional(map(list(string)), {})
-    service_iam_grants          = optional(list(string), [])
+    host_project                 = string
+    host_project_iam             = optional(list(string), [])
```
Using a list for type will make it very fragile as dynamic values in keys usually trigger an error on first apply, which was one of the reasons for not supporting this . :) We should mention it in the README (which you might already do, haven't gotten that far yet).

simonebruzzechesse replied:
Added warning for  that attribute! Thanks

ludoo commented on blueprints/factories/project-factory/variables.tf:
```diff
@@ -28,10 +28,12 @@ variable "data_defaults" {
+      host_project                 = string
+      host_project_iam             = optional(list(string), [])
+      service_identity_iam         = optional(map(list(string)), {})
+      service_identity_subnets_iam = optional(map(list(string)), {})
```
`service_identity_subnet_iam`

simonebruzzechesse replied:
Updated!

ludoo commented on blueprints/factories/project-factory/variables.tf:
```diff
@@ -28,10 +28,12 @@ variable "data_defaults" {
+      service_identity_iam         = optional(map(list(string)), {})
+      service_identity_subnets_iam = optional(map(list(string)), {})
+      service_iam_grants           = optional(list(string), [])
+      subnets_iam                  = optional(map(list(string)), {})
```
`subnet_iam`

simonebruzzechesse replied:
Replaced with network_subnet_users

ludoo commented on modules/project/README.md:
````diff
@@ -317,9 +317,53 @@ module "service-project" {
 # tftest modules=2 resources=9 inventory=shared-vpc-auto-grants.yaml e2e
 ```
 
-In specific cases it might make sense to selectively grant the `compute.networkUser` role for service identities at the subnet level, and while that is best done via org policies it's also supported by this module.
````
We should keep this sentence, as you're nuking the explanation for service identities. :)

simonebruzzechesse replied:
I just moved that sentence below the new example section

ludoo commented on modules/project/README.md:
```diff
@@ -317,9 +317,53 @@ module "service-project" {
+The 'networkUser' role for identities other than API services (e.g. users, groups or service accounts) can be managed via the `host_project_iam` attribute, by specifying the list of identities.
 
-In this example, Compute service identity will be granted compute.networkUser in the `gce` subnet defined in `europe-west1` region.
+Note that this configuration grants the 'networkUser' role at project level which results in the identity being able to configure resources on all the VPCs and subnets belonging to the host project without further restrictions. It is possible, or better recommended, to restrict which subnets can be used on the newly created project using an Org Policy. The latter applied at project level can rest ...
```
`The most reliable way to restrict which subnets [...] is via the xxxxx organization policy.`

ludoo replied:
~`The latter applied at project level can restrict access to a list of subnets from the host project.`~

simonebruzzechesse replied:
Updated!

ludoo commented on modules/project/README.md:
````diff
@@ -317,9 +317,53 @@ module "service-project" {
+# tftest modules=2 resources=11 inventory=shared-vpc-host-project-iam.yaml e2e
+```
+
+In specific cases it might make sense to selectively grant the `compute.networkUser` role for all kind of identities at the subnet level, although you can restrict access via org policies this is also supported by this module.
````
I would rephrase as

> In specific cases it might make sense to selectively grant the `compute.networkUser` role at the subnet level, and this is also supported by this module.

Then combine the following sentence in the same paragraph.

simonebruzzechesse replied:
Done!

ludoo commented on modules/project/README.md:
````diff
@@ -344,12 +388,66 @@ module "service-project" {
+```
+
+The last example provides a reference Shared VPC configuration that adheres to the principle of least privilege by:
+- Granting API services and other identities roles at the subnet level
````
I would just show the org policy in the project role example above and scrap this example, combining the two does not make a lot of sense IMHO.

simonebruzzechesse replied:
I added this last example just for showing a "reference" configuration for Shared VPC with permissions assigned at subnet level + org policy which might be a best practise, WDYT?

ludoo replied:
you don't need subnet-level with org policy IMHO :) I find this confusing, let's merge the org policy where we show project-level IAM

simonebruzzechesse commented on modules/project/variables.tf:
```diff
@@ -314,10 +316,12 @@ variable "shared_vpc_service_config" {
+      length(var.shared_vpc_service_config.subnets_iam) == 0
     )
-    error_message = "You need to provide host_project when providing service_identity_iam or service_iam_grants"
+    error_message = "You need to provide host_project when providing shared vpc host and subnets iam permissions."
```
Updated!

simonebruzzechesse commented:
> Great work Simo, thanks for this. Dropped a few comments mainly on variable naming and the README wording, once those are addressed this is good to go from my PoV.

Thanks for the feedback and comments, I updated the PR with your improvements! WRT what you suggested I just kept the reference example for the Shared VPC being, IMHO, the best configuration which suits most of the use cases. WDYT?

ludoo commented:
I don't think it's a reference as I wrote twice :) I would just move the org policy to the project-level IAM example, and scrap the last example.

ludoo reviewed (approved):
Thanks Simo!!!
